File: ios/MapsApp/Navigation.swift
import CoreLocation
import Foundation

/// Navigation maths, mirroring web/src/lib/navigation.ts so both platforms
/// behave identically. Coordinates are [lon, lat] to match route geometry.
enum Nav {
    static let earthRadius = 6_371_000.0
    /// Off-route threshold — generous enough for GPS noise and wide roads.
    static let offRouteM = 45.0

    static func distance(_ a: (lon: Double, lat: Double), _ b: (lon: Double, lat: Double)) -> Double {
        let lat = ((a.lat + b.lat) / 2) * .pi / 180
        let x = ((b.lon - a.lon) * .pi / 180) * cos(lat)
        let y = (b.lat - a.lat) * .pi / 180
        return (x * x + y * y).squareRoot() * earthRadius
    }

    /// Compass bearing a→b in degrees (0 = north) for the follow camera.
    static func bearing(_ a: (lon: Double, lat: Double), _ b: (lon: Double, lat: Double)) -> Double {
        let f1 = a.lat * .pi / 180, f2 = b.lat * .pi / 180
        let dl = (b.lon - a.lon) * .pi / 180
        let y = sin(dl) * cos(f2)
        let x = cos(f1) * sin(f2) - sin(f1) * cos(f2) * cos(dl)
        return (atan2(y, x) * 180 / .pi + 360).truncatingRemainder(dividingBy: 360)
    }

    private static func pointToSegment(
        _ p: (lon: Double, lat: Double),
        _ a: (lon: Double, lat: Double),
        _ b: (lon: Double, lat: Double)
    ) -> (d: Double, t: Double) {
        let scale = cos(p.lat * .pi / 180)
        let ax = a.lon * scale, ay = a.lat
        let bx = b.lon * scale, by = b.lat
        let px = p.lon * scale, py = p.lat
        let dx = bx - ax, dy = by - ay
        let len2 = dx * dx + dy * dy
        let t = len2 == 0 ? 0 : max(0, min(1, ((px - ax) * dx + (py - ay) * dy) / len2))
        let cx = ax + t * dx, cy = ay + t * dy
        let ddx = (px - cx) / scale, ddy = py - cy
        let d = ((ddx * scale) * (ddx * scale) + ddy * ddy).squareRoot() * (.pi / 180) * earthRadius
        return (d, t)
    }

    struct Snapped {
        var index: Int
        var offRouteM: Double
        var remainingM: Double
    }

    /// Snap a live position onto the route. Searching forward from the last
    /// index keeps this cheap and stops self-crossing routes from rewinding.
    static func snap(
        position: (lon: Double, lat: Double),
        geometry: [[Double]],
        from: Int = 0,
        window: Int = 400
    ) -> Snapped {
        guard geometry.count >= 2 else { return Snapped(index: 0, offRouteM: 0, remainingM: 0) }
        var best = (d: Double.infinity, i: from, t: 0.0)
        let end = min(geometry.count - 1, from + window)
        var i = max(0, from)
        while i < end {
            let a = (lon: geometry[i][0], lat: geometry[i][1])
            let b = (lon: geometry[i + 1][0], lat: geometry[i + 1][1])
            let r = pointToSegment(position, a, b)
            if r.d < best.d { best = (r.d, i, r.t) }
            i += 1
        }
        var remaining = distance(
            (geometry[best.i][0], geometry[best.i][1]),
            (geometry[best.i + 1][0], geometry[best.i + 1][1])
        ) * (1 - best.t)
        var j = best.i + 1
        while j < geometry.count - 1 {
            remaining += distance(
                (geometry[j][0], geometry[j][1]),
                (geometry[j + 1][0], geometry[j + 1][1])
            )
            j += 1
        }
        return Snapped(index: best.i, offRouteM: best.d, remainingM: remaining)
    }

    /// Which step is being driven, given progress along the geometry.
    static func currentStep(_ steps: [RouteStep], index: Int) -> Int {
        var s = 0
        for (i, step) in steps.enumerated() where step.beginIdx <= index { s = i }
        return s
    }

    /// Metres from the user to the start of the next maneuver.
    static func metresToManeuver(
        geometry: [[Double]], index: Int,
        position: (lon: Double, lat: Double), nextBeginIdx: Int
    ) -> Double {
        guard nextBeginIdx > index, geometry.count > 1 else { return 0 }
        let firstIdx = min(index + 1, geometry.count - 1)
        var m = distance(position, (geometry[firstIdx][0], geometry[firstIdx][1]))
        var i = index + 1
        while i < min(nextBeginIdx, geometry.count - 1) {
            m += distance((geometry[i][0], geometry[i][1]), (geometry[i + 1][0], geometry[i + 1][1]))
            i += 1
        }
        return m
    }
}

extension Array {
    /// Bounds-safe lookup for step indices that may momentarily run ahead.
    subscript(safe index: Int) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}

File: web/src/lib/navigation.ts
/// Navigation maths, kept free of React so it can be unit-reasoned about.
/// Everything works on [lon, lat] pairs, matching the route geometry.

export type LngLat = [number, number]

const R = 6_371_000

export function distanceM(a: LngLat, b: LngLat): number {
  const lat = ((a[1] + b[1]) / 2) * Math.PI / 180
  const x = ((b[0] - a[0]) * Math.PI / 180) * Math.cos(lat)
  const y = (b[1] - a[1]) * Math.PI / 180
  return Math.sqrt(x * x + y * y) * R
}

/** Compass bearing a→b in degrees (0 = north), for the follow camera. */
export function bearing(a: LngLat, b: LngLat): number {
  const φ1 = a[1] * Math.PI / 180, φ2 = b[1] * Math.PI / 180
  const Δλ = (b[0] - a[0]) * Math.PI / 180
  const y = Math.sin(Δλ) * Math.cos(φ2)
  const x = Math.cos(φ1) * Math.sin(φ2) - Math.sin(φ1) * Math.cos(φ2) * Math.cos(Δλ)
  return (Math.atan2(y, x) * 180 / Math.PI + 360) % 360
}

/** Perpendicular distance from p to segment a→b, plus the closest point. */
function pointToSegment(p: LngLat, a: LngLat, b: LngLat): { d: number; t: number } {
  const latScale = Math.cos((p[1] * Math.PI) / 180)
  const ax = a[0] * latScale, ay = a[1]
  const bx = b[0] * latScale, by = b[1]
  const px = p[0] * latScale, py = p[1]
  const dx = bx - ax, dy = by - ay
  const len2 = dx * dx + dy * dy
  const t = len2 === 0 ? 0 : Math.max(0, Math.min(1, ((px - ax) * dx + (py - ay) * dy) / len2))
  const cx = ax + t * dx, cy = ay + t * dy
  const ddx = (px - cx) / latScale, ddy = py - cy
  return { d: Math.sqrt((ddx * Math.cos((p[1] * Math.PI) / 180)) ** 2 + ddy * ddy) * (Math.PI / 180) * R, t }
}

export interface Snapped {
  /** Index of the geometry vertex the user is at or just passed. */
  index: number
  /** Metres from the route line — the off-route signal. */
  offRouteM: number
  /** Metres remaining to the end of the route. */
  remainingM: number
}

/// Snap a live position onto the route: which vertex we're at, how far off
/// the line we are, and how much is left. Searching from the last known
/// index forward keeps it O(window) and stops the route "rewinding" when a
/// path crosses itself.
export function snapToRoute(
  pos: LngLat,
  geometry: LngLat[],
  fromIndex = 0,
  window = 400,
): Snapped {
  if (geometry.length < 2) return { index: 0, offRouteM: 0, remainingM: 0 }
  let best = { d: Infinity, i: fromIndex, t: 0 }
  const end = Math.min(geometry.length - 1, fromIndex + window)
  for (let i = Math.max(0, fromIndex); i < end; i++) {
    const { d, t } = pointToSegment(pos, geometry[i], geometry[i + 1])
    if (d < best.d) best = { d, i, t }
  }
  // Remaining = rest of the current segment + every following segment.
  let remaining = distanceM(geometry[best.i], geometry[best.i + 1]) * (1 - best.t)
  for (let i = best.i + 1; i < geometry.length - 1; i++) {
    remaining += distanceM(geometry[i], geometry[i + 1])
  }
  return { index: best.i, offRouteM: best.d, remainingM: remaining }
}

/** The step being driven, given how far along the geometry we are. */
export function currentStep(steps: { beginIdx: number }[], index: number): number {
  let s = 0
  for (let i = 0; i < steps.length; i++) if (steps[i].beginIdx <= index) s = i
  return s
}

/** Metres from the user to the start of the next maneuver. */
export function metresToNextManeuver(
  geometry: LngLat[],
  index: number,
  pos: LngLat,
  nextBeginIdx: number,
): number {
  if (nextBeginIdx <= index) return 0
  let m = distanceM(pos, geometry[Math.min(index + 1, geometry.length - 1)])
  for (let i = index + 1; i < Math.min(nextBeginIdx, geometry.length - 1); i++) {
    m += distanceM(geometry[i], geometry[i + 1])
  }
  return m
}

/** Off-route threshold — generous enough for GPS noise and wide roads. */
export const OFF_ROUTE_M = 45
